File: src/classify/tree/LeafSimilarTrees.java
package classify.tree;

import others.tools.TreeNode;

import java.util.ArrayList;
import java.util.List;

public class LeafSimilarTrees {

    /**
     * 0ms 100.00%
     * 37MB 18.48%
     */
    public boolean leafSimilar(TreeNode root1, TreeNode root2) {
        List<Integer> leavesRoot1 = leaves(root1);
        List<Integer> leavesRoot2 = leaves(root2);

        if (leavesRoot1 == leavesRoot2) return true;
        if (leavesRoot1.size() != leavesRoot2.size()) return false;

        for (int i = 0; i < leavesRoot1.size(); i++) {
            if (!leavesRoot1.get(i).equals(leavesRoot2.get(i)))
                return false;
        }

        return true;
    }

    private List<Integer> leaves(TreeNode root) {
        List<Integer> res = new ArrayList<>();
        leavesHelper(res, root);

        return res;
    }

    private void leavesHelper(List<Integer> res, TreeNode root) {
        if (root == null) return;
        if (root.left == null && root.right == null)
            res.add(root.val);
        leavesHelper(res, root.left);
        leavesHelper(res, root.right);
    }
}

File: leetcode-scala/src/dfs/LeafSimilarTrees.scala
package dfs

import tools.TreeNode

import scala.collection.mutable

object LeafSimilarTrees {

  /**
   * 572 ms   100.00%
   * 52.9 MB  100.00%
   * @param root1
   * @param root2
   * @return
   */
  def leafSimilar(root1: TreeNode, root2: TreeNode): Boolean = {
    val leaves1 = leavesOfTree(root1)
    val leaves2 = leavesOfTree(root2)

    leaves1 sameElements leaves2
  }

  def leavesOfTree(root: TreeNode): Array[Int] = {
    val leaves = new mutable.ArrayBuffer[Int]()

    if (root.left == null && root.right == null) {
      leaves.addOne(root.value)
    } else {
      if (root.left != null) leaves.addAll(leavesOfTree(root.left))
      if (root.right != null) leaves.addAll(leavesOfTree(root.right))
    }

    leaves.toArray
  }

  def main(args: Array[String]): Unit = {
    val root1 = ConvertSortedArrayToBinarySearchTree.sortedArrayToBST(Array.range(0, 10))
    val root2 = ConvertSortedArrayToBinarySearchTree.sortedArrayToBST(Array.range(0, 10))
    val leaves1 = leavesOfTree(root1)
    val leaves2 = leavesOfTree(root2)


    leaves1.map(_ + " ").foreach(print)
    println()
    leaves2.map(_ + " ").foreach(print)
  }
}
